Add cart alteration methods

adamvoncorswant opened the pull request:
## Cart: Adds `getId()` to item interface
So that refund logic can easier match what items are being returned

## Cart: Add functionality for item line alterations
This functionality can be used when doing partial refunds (or adding to
cart). Main methods to be used on a cart for this purpose is:

`Cart::getAlterableItems(previousAlterations)`
`Cart::getAlterationValue(previousAlterations, alteration)`

Intentional limitations:
* Will not create a new 'cart' representing a partial refund, because
  that would be confusing and logically inconsistent

Also adds
* new 'temp'-implementations of the three interfaces. This is to
  be able to modify quantities without altering the public interface: the
  classes are not exposed publicly anyway.
* `quantity` to ServiceCharge for consistency, so it behaves in
  the same way as `Discount`, and can alter direction in the same way

This is WIP. Thoughts on this approach @calle , @jakanjakan and @gaiazettle ?

adamvoncorswant commented on izettle-cart/src/main/java/com/izettle/cart/Item.java:
```diff
@@ -40,4 +40,11 @@
+     * @param <I> the type of the identifier, can be of any type as long as it's comparable
+     * @return the identifier of the item
+     */
+    <I extends Object> Comparable<I> getId();
```
I'm not sure if it's worth having a generic type for the `id` member. An alternative would be to use UUID directly, but then we'd have a problem with old carts where we (today) use a combination of other fields for comparison.

adamvoncorswant commented on izettle-cart/src/test/java/com/izettle/cart/AlterationTest.java:
```diff
@@ -0,0 +1,140 @@
+        );
+        assertEquals(originalCart.getValue(), 5L);
+        final List<Map<UUID, BigDecimal>> previousAlterations = new ArrayList<Map<UUID, BigDecimal>>();
+        final int[] expectedValues = {1, 0, 1, 0, 1, 0, 1, 0, 1, 0};
```
@gaiazettle : these expected values correspond to the scenario discussed IRL, where today you'd get the same value every time.

xiaodong-izettle commented:
@adamvoncorswant, I added a commit to make the Generics usage look a little bit better, and remove a usage of a method introduces in Java 1.8, hope you do not mind... 😄

xiaodong-izettle commented on izettle-cart/src/main/java/com/izettle/cart/AlterationUtils.java:
```diff
@@ -54,20 +54,27 @@ void validateItems(
         return mergedAlterations;
     }
+
+    private static <I extends Comparable<I>, V> V getOrDefault(Map<I, V> map, Object key, V defaultValue) {
```
Generally copied the code from JDK...

adamvoncorswant replied:
Ah, thanks. Indeed weird that this did not cause any compilation issue. Might be worth looking into!

xiaodong-izettle replied:
Yeah, agree... I'll look into that... 😄

xiaodong-izettle commented on izettle-cart/src/main/java/com/izettle/cart/AlterationUtils.java:
```diff
@@ -0,0 +1,73 @@
+                for (Entry<I, BigDecimal> entry : previousAlteration.entrySet()) {
+                    final I id = entry.getKey();
+                    final BigDecimal quantity = entry.getValue();
+                    mergedAlterations.put(id, mergedAlterations.getOrDefault(id, BigDecimal.ZERO).add(quantity));
```
Weird that this did not trigger any build failure... `Map::getOrDefault` was introduced in `1.8`, and `izettle-cart` should be compiled with `1.6`...

adamvoncorswant commented on izettle-cart/src/main/java/com/izettle/cart/Item.java:
```diff
@@ -46,5 +46,5 @@
      * @return the identifier of the item
      */
-    <I extends Object> Comparable<I> getId();
+    <I extends Comparable<I>> Comparable<I> getId();
```
Nice. Just saw that mistake locally as well! I was changing back and forwards between the two, and must have forgotten to change the interface in the end ;)

xiaodong-izettle commented on izettle-cart/src/main/java/com/izettle/cart/AlterationUtils.java:
```diff
@@ -64,17 +66,10 @@ void validateItems(
                     final I id = entry.getKey();
                     final BigDecimal quantity = entry.getValue();
-                    mergedAlterations.put(id, getOrDefault(mergedAlterations, id, BigDecimal.ZERO).add(quantity));
+                    mergedAlterations.put(id, coalesce(mergedAlterations.get(id), BigDecimal.ZERO).add(quantity));
```
Hehe, cool... 👍

adamvoncorswant replied:
:smile: I thought it was a bit overkill with a specific method for this tiny case. Hope you didn't mind.

xiaodong-izettle replied:
Yeah, sure... 😄

adamvoncorswant commented:
1. A alteration is not done using information from the product library, but only the original cart (and possible other refunds). The itemline being altered need not even be in the product library at the moment of the refund. 
2. Because of that, no other information than identifier and quantities are needed to perform a refund/alteration
3. The alteration will *always* need it's original cart (and possible other alterations) to make sense, or to even have a value: in itself the alteration does not carry enough information

What I don't really have a solution for right now is how one should store an alteration:
* As a "fake" cart with itemlines, discounts and service charges. This is nice, because it doesn't alter current storage models (where refunds and normal purchases are stored in the same way). The drawback is that the information cannot be passed into cart logic "as-is", because it would then possibly get the wrong value.
* As something different, with just the identifiers and quantities, possibly with the value explicitly saved for clarity. Drawback here is that this alters storage model and will be more complex to roll out. Possibly even in toolbox-cart, as a "refund" rather than a "cart" object?

Please take your time and think about this. I'd love to hear your thoughts about this @calle , @jakanjakan and @gaiazettle . It's also entirely possible that I've overthought this over the holidays :santa:

calle commented on izettle-cart/src/main/java/com/izettle/cart/Item.java:
```diff
@@ -40,4 +40,11 @@
+     * @param <I> the type of the identifier, can be of any type as long as it's comparable
+     * @return the identifier of the item
+     */
+    <I extends Comparable<I>> Comparable<I> getId();
```
Lovely! 

How about moving the definition of `I` to the `Item` interface as for `T` and `K`?

adamvoncorswant replied:
I think I agree, given that we want to specify this using generics to begin with. Or do you see any other way of enforcing this method without specifying it's type (don't even see that `Comparable` is very important here) ?

calle commented on izettle-cart/src/main/java/com/izettle/cart/ItemUtils.java:
```diff
@@ -68,4 +68,19 @@ private static BigDecimal getExactGrossValue(final Item item) {
+            throw new IllegalArgumentException("Quantity cannot be null");
+        }
+        if (quantity.compareTo(BigDecimal.ZERO) == 0) {
+            throw new IllegalArgumentException("Quantity cannot be ZERO: " + quantity);
```
This might be a bit to generic? Should we never allow 0 quantities? I'm thinking of a cart with fully refunded items or other "marker" line items as we discussed earlier?

adamvoncorswant replied:
Saw this now, 10 months later :) Well, this is specifically used only for alterations, and I'd say that altering 0 nr of items is not an alteration...

calle commented on izettle-cart/src/main/java/com/izettle/cart/ServiceCharge.java:
```diff
@@ -1,8 +1,16 @@
+     * The quantity, or the number of units that this item represents
+     * @return the quantity, cannot be null
+     */
+    BigDecimal getQuantity();
```
👍  for consistency!

calle commented on izettle-cart/src/main/java/com/izettle/cart/TempDiscount.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Concrete implementation of Discount interface used for internal temporary calculations: never exposed publicly
+ */
+class TempDiscount implements Discount<TempDiscount> {
```
No really fond of these names. `Temp` sounds like it is something temporary in terms of the code being added as a temporary fix later to be removed.

Not sure what a better naming would be though. Maybe better to create an explicit `PartialCart` that has internal implementations of these internally?

calle replied:
Maybe even calling it `AlteredCart` since it is only (for now) used for alterations?

adamvoncorswant replied:
Sounds good! I'll change!

calle commented on izettle-cart/src/main/java/com/izettle/cart/AlterationUtils.java:
```diff
@@ -0,0 +1,75 @@
+        for (ItemLine<T, D> itemLine : cart.getItemLines()) {
+            final T item = itemLine.getItem();
+            final Comparable<I> id = item.getId();
+            final BigDecimal quantity = CartUtils.coalesce(quantityById.get(id), BigDecimal.ZERO);
```
Use static import of `CartUtils.coalesce`, maybe even align the code with the similar code in `mergeAlterations` below.

adamvoncorswant replied:
Check

calle commented on izettle-cart/src/main/java/com/izettle/cart/Cart.java:
```diff
@@ -47,8 +52,142 @@ public Cart(final List<T> items, final List<K> discounts, final S serviceCharge)
+            final Map<I, BigDecimal> alteration
+    ) {
+        ItemUtils.validateQuantities(alteration.values());
+        //Verify that all referenced items are present in the original cart
```
Should we validate that the alteration values are negative?

calle replied:
Strike that, I now see your code is generic to handle "top-ups" with positive quantities as well.

adamvoncorswant replied:
Check

calle commented on izettle-cart/src/main/java/com/izettle/cart/Cart.java:
```diff
@@ -47,8 +52,142 @@ public Cart(final List<T> items, final List<K> discounts, final S serviceCharge)
+        for (ItemLine<T, D> itemLine : this.getItemLines()) {
+            final T originalItem = itemLine.getItem();
+            TempItem newItem = null;
+            for (Map.Entry<I, BigDecimal> entry : alteration.entrySet()) {
```
Why not `alteration.contains(originalItem.getId())`?

adamvoncorswant replied:
You're right of course. I think this was altered/tested in so many versions so I lost track :)

calle commented on izettle-cart/src/main/java/com/izettle/cart/Cart.java:
```diff
@@ -47,8 +52,142 @@ public Cart(final List<T> items, final List<K> discounts, final S serviceCharge)
+            if (newItem == null) {
+                newItem = TempItem.from(originalItem);
+            }
+            if (newItem.getQuantity().compareTo(BigDecimal.ZERO) != 0) {
```
Why this check? Should we not keep the items with quantity 0? Or is this because they are not used for the calculations done in this class?

adamvoncorswant replied:
The thought was that, as the items were completely drained from the original cart, we don't need that specific line anymore. Do you think we should keep it anyway?

calle commented on izettle-cart/src/main/java/com/izettle/cart/Cart.java:
```diff
@@ -47,8 +52,142 @@ public Cart(final List<T> items, final List<K> discounts, final S serviceCharge)
+            = cartAfterPreviousAlterations.applyAlteration(alteration);
+        final long valueAfterPreviousAlterations = cartAfterPreviousAlterations.getValue();
+        final long valueAfterNewAlteration = afterLatestAlteration.getValue();
+        return valueAfterPreviousAlterations - valueAfterNewAlteration;
```
I get a bit confused by the naming here. `cartBeforeLastAlteration` and `cartAfterLatestAlteration` as well as `valueBeforeLastAlteration` and `valueAfterLastAlteration` is clearer I think.

adamvoncorswant replied:
Sounds good!

calle commented on izettle-cart/src/main/java/com/izettle/cart/Cart.java:
```diff
@@ -47,8 +52,142 @@ public Cart(final List<T> items, final List<K> discounts, final S serviceCharge)
+     * @param previousAlterations possibly previous alterations that needs to be taken into consideration
+     * @return A map of quantities available for alteration
+    */
+    public <I extends Comparable<I>> Map<Comparable<?>, BigDecimal> getAlterableItems(
```
Is it too hard to get this to return `Map<I, BigDecimal>`?

adamvoncorswant replied:
Well, the problem is that the generic type parameter `I` has two different values, depending on if it's from an `Item` or an `CartAlterationItem`... I guess that it's basically a bad idea trying to genericise a single method on an interface that is otherwise genericised on a class level.

The reason why I chose/tried to do it this way was because I felt that the syntax required just to declare a cart (elsewhere in calling code) was reaching absurd levels of generic bonanza. 
What's your thought on that? I see three options:
* Declare `public Object getId();` and accept the ugliness of that.
* Make the `I` type declared on a class level (as per your suggestion), and live with the verbose expressions this results in.
* Simply say that this will be `UUID`s at all times, and declare `public UUID getId();`.

adamvoncorswant commented:
Thanks for great feedback so far! I've amended most of the comments, except for the ones related to the type parameter `I`, that needs more discussion.

adamvoncorswant commented:
I've added the type parameter `I` to the item interface in another branch. For comparison, please see: https://github.com/iZettle/izettle-toolbox/compare/add-cart-alteration-methods...add-cart-alteration-methods-2

calle commented:
I see your point about "absurd levels of generic bonanza" 😄  with `Cart<T, D, K, S, I>`...

We could of course extend the Cart class in the various services to provide a "typed" Cart for the service in question. For example a `public class PurcahseCart extends Cart<Product, ProductDiscount, Discount, ServiceCharge> {...}` but it will probably affect a lot of code.

From your three alternatives my comments are:

1. `Object getId()`: I kind of like it, the identifier could really be anything and the only requirement is that we could perform an `equals()` (with a consistent `hashCode()`) on it just like Object. It also allows any implementing classes to further narrow down the type of the identifier if needed:
```java
public class Product implements Item<Product, ProductDiscount> {
    ...
    private UUID productUuid;
    ...
    @Override
    public UUID getId() {
        return productUuid;
    }
    ...
}
```

2. Add `I` to `Item` interface: Will be verbose and not sure we need it as per above. Could be argued to be the "most correct" way of doing this.

3. Forcing `UUID`: Not really fond of this one even though we probably will return `UUID` in all cases. If feels a bit too limiting without any real benefit.

Was there a specific reason for having `I extends Comparable<I>`?

adamvoncorswant commented:
Our IRL discussion leaned towards not doing this, and live with rounding problems when it comes to multiple refunds.

Let's leave it around for a while, until we've decided!

calle commented:
Has the "for a while" passed now perhaps?

adamvoncorswant commented:
8 months? You think so? Hehe, fair enough. Feels bad though, as this would be the logically correct way of doing it :cry:

adamvoncorswant commented:
Revitalizing this one again, as it seems it might have some merits after all. Let's discuss this during next week @calle , @gaiazettle and @vincenthauser

calle commented:
> Our IRL discussion leaned towards not doing this, and live with rounding problems when it comes to multiple refunds.

I don't really remember what the arguments against this were, do you? In the PR it is mostly on the design of the code but I have a vague memory there was something else stopping us from going down this route...

adamvoncorswant commented:
I think it boiled down to the cost of changing existing implementations (storage model for refund purchases needs to be changed if 100% correctness is to be achieved). However, @vincenthauser is already working on changing some of the Android refund logic, and I believe that at least providing the tool for doing it correctly might increase our chances to do it across the board eventually ;)

adamvoncorswant commented:
Added some documentation in README, also added another test.

adamvoncorswant commented:
What I'm curious about here is feedback on the `public` methods on the cart `Cart::getAlterationValue` and `Cart::getAlterableItems`. Do they (and their signatures) make sense, and would they be of use for a client developer @vincenthauser ?

adamvoncorswant commented:
TODO:
* Return something more complex than just a `long` as the client will need VAT etc
* Add test to verify that line item discounts behave as expected.

adamvoncorswant commented:
Item line discounts now behave as expected.

TODO:
* Return something more complex than just a long as the client will need VAT etc

adamvoncorswant commented:
fyi @MrShish

adamvoncorswant commented:
Chose to rename the public methods to something `refund` instead of `alteration` to make it easier for outside callers. Also added a `RefundCart` object that is returned to the caller, where all basic data can be retrieved.

@vincenthauser is now trying this out on a snapshot build for integration in his project.

adamvoncorswant commented:
Ok, I'm done for now. Me and @vincenthauser have been pair coding usage of this functionality in the Android app with great success. 

Squashed, rebased and force pushed.

Anyone up for merging this?

adamvoncorswant commented:
@calle , @gaiazettle or @MrShish : you up for merging this now?

MrShish commented on izettle-cart/src/main/java/com/izettle/cart/Cart.java:
```diff
@@ -47,8 +53,151 @@ public Cart(final List<T> items, final List<K> discounts, final S serviceCharge)
+     * @return the value of the alteration
+     */
+    public AlterationCart<T, D, K, S> createAlterationCart(
+        final List<Map<UUID, BigDecimal>> previousAlterations,
```
Sorry gotta ask. How do we keep track of the previous alterations?
So if:

1. A purchase is made. Say 3 shoes
2. A partial refund is made. Returning back 1 Shoe
3. Another partial refund is made, another shoe is returned.

From reading the tests, the alteration is the delta between the steps: (id, -1)
To create a refund, we need to load the purchase, and all previous refunds and treat the refunds as alterations?

adamvoncorswant replied:
Exactly so. So you cannot know the value of an alteration until you have the full history as well. While I understand that this looks a bit odd, I've come to the conclusion that this is the only way to do it. Good thing is that it doesn't seem to be any problems for our apps (as they need to request the full history anyway so that the GUI can show what items are refundable). Bad thing is that some of our other services will have to change a bit for this to work.

However, using this functionality is optional, and meant as a tool to be able to 'do the right thing' ™. As long as the backend services doesn't require all the numbers from the client to be exactly right down to the least denomination, it's fine to keep it as it is.

MrShish replied:
check!

MrShish commented on izettle-cart/src/main/java/com/izettle/cart/Item.java:
```diff
@@ -40,4 +41,10 @@
+     * The identifier of the item in relation to other line items in the cart
+     * @return the identifier of the item
+     */
+    UUID getId();
```
Does this need to be UUID? Aren't we forcing the user of this utility to use only UUIDs? 
Some might use Longs, or Strings as identifier.. no?

adamvoncorswant replied:
I know. The thing is that I've tried with three other approaches and I didn't like the outcome of either of them:
1. Genericize the id type, making all things beautiful except for the generic bonanza in the declaration of the cart (which will have to be written out by all clients instantiating such an object)
2. Local generics (outside of the cart declaration) also works quite well, but then there will be lots of type warnings in the code, as there is nothing ensuring that the id types from the carts items matches the ones sent in to it's alteration method by am outside caller.
3. Just go with `Object`. This was probably the least evil one, although it felt so 90:ies to pass around an `Object` everywhere. Also, as java collections doesn't support subtypes of a generically created collection, `UUID` would have to be cast to `Objects` in quite a lot of calling code.

So, I just thought to myself: screw it. We're only ever gonna use UUID anyway, so why go through all the trouble?

MrShish replied:
thanks for the explanation, I trust your judgement :)

adamvoncorswant replied:
:heart:

adamvoncorswant commented on izettle-cart/src/main/java/com/izettle/cart/AlteredCartItem.java:
```diff
@@ -0,0 +1,96 @@
+    private final Float vatPercentage;
+    private final BigDecimal quantity;
+    private final AlteredCartDiscount discount;
+    private final UUID id;
```
Hmm. Yes, it will. Would it be problematic to create one on the fly for those products?

MrShish replied:
I guess not, thou im not found of creating ids, that wont exist anywhere. Thinking about purchases that has on the fly products, with uuids, but doesn't exists in product library.. 
Can get confusing.

adamvoncorswant replied:
I kind of agree. The only other alternative would be to assign cart internal identifiers to all item lines (eg 0,1,2) and pass those back to the caller when indicating how many of each item is refundable.
Then, the caller gets back with a reference to this number instead of a UUID... 
Problem with this approach is that the internal number is relative and not persistent. How would you then persist your refund: by persisting the relative index? Then you *really* would always need the full history each time you even look at a refund to understand what it's about.

Also, the above approach probably becomes even more problematic the day we want to introduce additions of new products (exchange).

MrShish replied:
I understand, but this means that this util, essentially affects the business-logic/architecture of how we handle products . I know its not a biggie to generate uuids on the fly, but. Conceptually, this is the case. 
I don't see any good alternative solution either, except maybe doing a composite key and using it as the identifier of the product (name,variant,vat,price,unit). Then it would be up to the caller to decide what the id is, and where back to Object :/

adamvoncorswant replied:
So you're basically suggesting that I undo the last commit? I can live with that!

adamvoncorswant replied:
If you promise to merge after I force push?! Otherwise @vincenthauser will strangle me (as I've had him switch key type a few times now) :scream_cat:

MrShish replied:
haha, i dont want the wrath of the hauser. I promise!

adamvoncorswant commented on izettle-cart/src/main/java/com/izettle/cart/AlteredCartItem.java:
```diff
@@ -24,7 +25,7 @@
 
     private AlteredCartItem(
-        final Object id,
+        final UUID id,
```
No, just `Item`s (eg, not discounts or service charge)

MrShish commented:
Only had some minor comments @adamvoncorswant ! Impressive work! 
So fun that a thing you think is trivial is actually quite complicated. Imagine having these calculations sprinkled across all of our services.. 
Now if we can port this to swift 🤔

adamvoncorswant commented on izettle-cart/src/main/java/com/izettle/cart/AlterationCart.java:
```diff
@@ -0,0 +1,108 @@
+ * @param <K> the type of the cart-wide discounts
+ * @param <S> the type of the service charge
+*/
+public class AlterationCart<T extends Item<T, D>, D extends Discount<D>, K extends Discount<K>, S extends ServiceCharge<S>> {
```
:laughing: Yes, and this is just the same as the generics for `Cart`. Would we have included the type of the identifier (eg UUID), we would have added one more :)

FredrikMunter replied:
I stumbled upon this beautiful use of generics yesterday by ways of an accidental click. Honestly, the most confusing thing I ever saw. Hat's off,good sir! 🎉 

![image](https://user-images.githubusercontent.com/1296440/33162350-747d9838-d029-11e7-9d4f-ab7b48bcfddf.png)

adamvoncorswant replied:
Well thank you friend! We're trying hard to come up with something using even more confusing generics, but it seems almost impossible. 😂

adamvoncorswant commented on izettle-cart/src/test/java/com/izettle/cart/AlterationTest.java:
```diff
@@ -0,0 +1,228 @@
+        );
+        assertEquals(originalCart.getValue(), 5L);
+        final List<Map<UUID, BigDecimal>> previousAlterations = new ArrayList<Map<UUID, BigDecimal>>();
+        final int[] expectedValues = {-1, 0, -1, 0, -1, 0, -1, 0, -1, 0};
```
I think this test is one of the more explanatory ones. It verifies that returning the same kind of item again and again will eventually drain the cart completely (so that all money is paid back), and that each return in itself might yield a different value than the previous because of rounding phenomenon.

adamvoncorswant commented:
Done. We're back to `Object` as identifier, and it's up to the user of the library to use UUIDs or object hashes or whatever. 

Rebased and force pushed!

adamvoncorswant commented:
Woho! The day has come! 

Thanks @MrShish (and the rest)

MrShish commented:
![4415631-1runyah](https://user-images.githubusercontent.com/108077/31469798-7792a7b4-aee3-11e7-9852-4623795b580d.jpg)

😂

vincenthauser commented:
I saw that!

On 12 October 2017 at 00:16, Umut Yurdakök <notifications@github.com> wrote:

> *@MrShish* commented on this pull request.
> ------------------------------
>
> In izettle-cart/src/main/java/com/izettle/cart/AlteredCartItem.java
> <https://github.com/iZettle/izettle-toolbox/pull/205#discussion_r144150826>
> :
>
> > @@ -0,0 +1,96 @@
> +package com.izettle.cart;
> +
> +import java.math.BigDecimal;
> +import java.util.UUID;
> +
> +/**
> + * Concrete implementation of Item interface used for internal temporary calculations: never exposed publicly
> + */
> +class AlteredCartItem implements Item<AlteredCartItem, AlteredCartDiscount> {
> +
> +    private final long unitPrice;
> +    private final Float vatPercentage;
> +    private final BigDecimal quantity;
> +    private final AlteredCartDiscount discount;
> +    private final UUID id;
>
> haha, i dont want the wrath of the hauser. I promise!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/iZettle/izettle-toolbox/pull/205#discussion_r144150826>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGxU-8Nt_jDncBRptTVnh1QBe3Wc-3mzks5srT5ZgaJpZM4LY60r>
> .
>



-- 

––
Vincent Hauser
Software Developer
iZettle

+46 707-40 10 25
www.izettle.com <http://izettle.com/>
